src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_hash().is_some() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        if !check_pin_auth(&shared_secret, &new_pin_enc, &pin_auth) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        if !check_and_store_new_pin(persistent_store, &aes_dec_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        persistent_store.reset_pin_retries();
+        Ok(())
+    }
+
+    fn process_change_pin(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let mut auth_param_data = new_pin_enc.clone();
+        auth_param_data.extend(&pin_hash_enc);
+        if !check_pin_auth(&shared_secret, &auth_param_data, &pin_auth) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
+
+        if !check_and_store_new_pin(persistent_store, &aes_dec_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        self.pin_uv_auth_token = rng.gen_uniform_u8x32();
+        Ok(())
+    }
+
+    fn process_get_pin_token(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
```

Comment:
This part is similar as well (though not exactly the same). Maybe some utility function to "exchange and check shared secret"?

Reply:
Part of the problem is that I am often not allowed to reorder instructions - for code logic or protocol reasons. I could make this a function

`        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);`
`        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);`
`        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;`

That outputs the AES keys is a result, but doesn't seem worth the extra level of indirection.